Multi-dot globbing

quasicomputational opened the pull request:
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/#conventions).
* [x] Any changes that could be relevant to users have been recorded in the changelog.
* [x] The documentation has been updated, if necessary.
* [x] If the change is docs-only, `[ci skip]` is used to avoid triggering the build bots.

Please also shortly describe how you tested your change. Bonus points for added tests!

New test added for the `cabal check` and specific tests for this behaviour have been added to the pre-existing glob tests.

The output of `cabal check` has changed a bit, but not for the worse, I don't think; it's now more like the other commands.

23Skidoo commented on Cabal/Distribution/Simple/Glob.hs:
```diff
@@ -86,35 +106,59 @@ explainGlobSyntaxError filepath VersionDoesNotSupportGlob =
       let (candidateBase, candidateExts) = splitExtensions (last $ seg:segs)
-      in ext == candidateExts && not (null candidateBase)
-    FinalMatch NonRecursive ext ->
+      guard (not (null candidateBase))
```
So `*.html` won't match `.foo.html`. Makes sense, since that's how it works in shell, but we should document it.

quasicomputational commented:
Thanks! Rebased and squashed for tidiness and should be ready to merge, but I'll let it sit for a few days to see if anyone else spots anything that needs changing before the big green button gets pressed.
